### ISSUE TITLE
Module X has no attribute Y when using star imports

### ISSUE THREAD

**Issue by nicktimko** [python/mypy#4930]
mypy seems to emit a spurious warning that it can't find an attribute of a module that should be imported via a star import and where another import refers to it.

## Input

`$ head mypackage/*`
* `==> mypackage/__init__.py <==`
```python
from mypackage.submod1 import *     # * -> HELLO and error goes away
from mypackage.submod2 import greet # comment out and error goes away
__all__ = ['HELLO', 'greet'] # does nothing
```

* `==> mypackage/submod1.py <==`
```python
__all__ = ['HELLO']
HELLO = 'hello there'
```

* `==> mypackage/submod2.py <==`
```python
from mypackage import HELLO
__all__ = ['greet']
def greet():
    print(HELLO)
```

## Current Output

```bash
$ mypy mypackage
mypackage/submod2.py:1: error: Module 'mypackage' has no attribute 'HELLO'
$ python -c "import mypackage;mypackage.greet()" # Python has no problems
hello there
$ python --version; mypy --version
Python 3.6.4
mypy 0.590
```

(haven't checked with master, but I couldn't find any recently closed issues that seemed to talk about something like this)

## Expected Output

No error from mypy.

As per the comments in the `__init__.py`, if some of the imports are tweaked, then mypy is happy. It doesn't seem to be sated by an `__all__` to help the star import. I don't quite understand why not importing `mypackage.submod2` causes the error to go away as well; it's not like there's a circular import between 1 and 2 where the names in the module aren't defined before it imports (i.e. if you moved the `...import greet` line above the other, Python would choke)

**Comment by jacebrowning** [comment]
I can still reproduce this with `mypy 0.600`.

**Comment by JelleZijlstra** [comment]
I looked into this for a bit. Here is a test that reproduces the problem: https://github.com/JelleZijlstra/mypy/commit/6b124e2578e73cad80444833ce0b43ab481a40c8. The problem goes away when one uses `from b import HELLO` instead of `from b import *`.

I think what's going on here is as follows:
- semantic analyzer pass 1 adds names from `import_from`, but not `import_alll` (because at that point we don't know anything about other modules yet)
- semantic analyzer pass 2 adds names from `import_all`, relying on what was added in pass 1. But if we're trying to import from a module that uses import_all, and pass 2 hasn't run on that module yet, we can't see the names from the import_all, and we get an error.

This could be fixable with an additional semanal pass or something like a callback in pass 1 that runs when the module we're import_alling from is processed, and adds names to the module that is using import_all.

**Comment by JelleZijlstra** [comment]
My branch https://github.com/python/mypy/compare/master...JelleZijlstra:circularstar has a fix for the original bug now, but it introduces various new bugs. I don't understand for now what's going on in some of those tests. I'll take another look at this later.

**Comment by sebastian-philipp** [comment]
not supporting `import *` is a bit unfortunate. :disappointed:. I guess we can live with it. Needs proper documentation, though.

**Comment by gvanrossum** [comment]
Clarification: we'd document that we don't support `import *` *in an import
cycle*. It seems expendable, since you can add an explicit import where
necessary.

**Comment by sebastian-philipp** [comment]
ok, the effect I'm seeing here is then caused by something else. thanks!

<details>
  <summary>Click to expand</summary>

```bash
    #  cat <<EOF > ./mypy.ini
    [mypy]
    strict_optional = True
    no_implicit_optional = True
    ignore_missing_imports = True
    warn_incomplete_stub = True
    check_untyped_defs = True
    show_error_context = True
    EOF
    #  cat test_foo.py 
    from nose.tools import *
    def test_foo():
        eq_(1,1)
    #  nosetests test_foo.py 
    .
    ----------------------------------------------------------------------
    Ran 1 test in 0.000s
    OK
    #  mypy --config-file=mypy.ini test_foo.py
    test_foo.py: note: In function "test_foo":
    test_foo.py:3: error: Name 'eq_' is not defined
```

</details>

**Comment by JukkaL** [comment]
This might be fixed by the new semantic analyzer (experimental version available on GitHub mypy master by using `--new-semantic-analyzer`). It generally improves the processing of imports within import cycles.

**Comment by chadrik** [comment]
I have the same issue.  I tried with `0.710+dev.f0491839adf2db2da74768b1d01f8e244e769f59` using `--new-semantic-analyzer` and it did not fix the problem.

**Comment by chadrik** [comment]
If you do decide to drop support for `import *` in an import cycle, it would be great to at least inform the user about the source of the cycle.  In my case it's not obvious, and it could be (???) caused by imports inside a `TYPE_CHECKING` block, which I can't remove without invalidating my annotations.

I would really rather mypy did not drop support for `import *` in a cycle.

**Comment by JukkaL** [comment]
We are not planning to drop support for `import *` in import cycles, but there may be specific use cases that are too hard to support.

**Comment by JukkaL** [comment]
I'm dropping the new-semantic-analyzer label since this doesn't seem specific to it.

**Comment by nicktimko** [comment]
Would it be any easier to support it where the imported module/package defines a (static) `__all__`?

**Comment by marmistrz** [comment]
This happens while using pygame: 
```
from pygame.locals import KEYUP
```
results in 
```
test.py:1: error: Module 'pygame.locals' has no attribute 'KEYUP'
```

**Comment by bczsalba** [comment]
Any update on this? If anything, there *really* should be something to tell the user what the issue is, as I just spent 20 minutes trying to figure out the cause.

Though, best would be if `from module import *` would be supported, at least when `__all__` is defined. For now you have to either manually import everything or manually re-type every field, even though the information is there.

**Comment by ilevkivskyi** [comment]
FWIW the original example passes cleanly on current master (both with and without `__all__`). So closing this for now.
